File: go-wasm/pkg/wasm/claimer.go
// +build wasm

package wasm

import (
	"encoding/hex"
	"encoding/json"
	"errors"
	"syscall/js"

	"github.com/KILTprotocol/portablegabi/go-wasm/pkg/credentials"
	"github.com/privacybydesign/gabi"
	"github.com/privacybydesign/gabi/revocation"
)

// GenKey creates the private key for the claimer
func GenKey(this js.Value, inputs []js.Value) (interface{}, error) {
	keyLength := DefaultKeyLength
	if len(inputs) > 0 && !inputs[0].IsUndefined() {
		keyLength = inputs[0].Int()
	}
	sysParams, success := gabi.DefaultSystemParameters[keyLength]
	if !success {
		return nil, errors.New("invalid key length")
	}

	claimer, err := credentials.NewClaimer(sysParams)
	if err != nil {
		return nil, err
	}
	return claimer, nil
}

// KeyFromSeed derives a key from a given seed
func KeyFromSeed(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 1 {
		return nil, errors.New("missing seed to generate claimer keys")
	}
	// get seed
	hexString := inputs[0].String()
	if len(hexString) < 4 || hexString[:2] != "0x" {
		return nil, errors.New("seed should be a hexadecimal string starting with '0x' followed by at least two hexadecimal digits")
	}
	seed, err := hex.DecodeString(hexString[2:])

	// get optional key length
	keyLength := DefaultKeyLength
	if len(inputs) > 2 && !inputs[2].IsUndefined() {
		keyLength = inputs[2].Int()
	}
	sysParams, success := gabi.DefaultSystemParameters[keyLength]
	if !success {
		return nil, errors.New("invalid key length")
	}

	// create claimer
	claimer, err := credentials.NewClaimerFromBytes(sysParams, seed)
	if err != nil {
		return nil, err
	}
	return claimer, nil
}

// RequestAttestation creates a session object and a message which request the
// attestation of specific attributes. The second object should be sent to an
// attester. This method expects as inputs the private key of the claimer, a
// json encoded string containing the claim which should be attested, the
// handshake message from the attester and the public key of the attester.
func RequestAttestation(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 4 {
		return nil, errors.New("missing inputs to request attestation")
	}
	claimer := &credentials.Claimer{}
	claim := credentials.Claim{}
	handshakeMsg := &credentials.StartSessionMsg{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), &claim); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), handshakeMsg); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[3].String()), issuerPubKey); err != nil {
		return nil, err
	}

	session, msg, err := claimer.RequestAttestationForClaim(issuerPubKey, handshakeMsg, claim)
	if err != nil {
		return nil, err
	}
	return map[string]interface{}{
		"session": session,
		"message": msg,
	}, nil
}

// BuildCredential creates a credential which can be used to create proofs that
// the claimer posseses certain attributes. This method expects as input the
// private key of the claimer, the session object created by requestAttestation
// and the signature message send the attester.
func BuildCredential(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 3 {
		return nil, errors.New("missing inputs to build credential")
	}
	claimer := &credentials.Claimer{}
	session := &credentials.UserIssuanceSession{}
	signature := &gabi.IssueSignatureMessage{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), session); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), signature); err != nil {
		return nil, err
	}

	credential, err := claimer.BuildCredential(signature, session)
	if err != nil {
		return nil, err
	}
	return credential, nil
}

// BuildPresentation creates a proof that the claimer posseses the requested
// attributes. This method takes as input the private key of the claimer, the
// credential which contains the requested attributes, a json encoded list
// containing the requested attributes and the public key of the attester.
// It returns a proof containing the values of the requested attributes.
func BuildPresentation(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 4 {
		return nil, errors.New("missing inputs to build presentation")
	}
	claimer := &credentials.Claimer{}
	credential := &credentials.AttestedClaim{}
	request := &credentials.PresentationRequest{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), credential); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), request); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[3].String()), issuerPubKey); err != nil {
		return nil, err
	}

	disclosedAttr, err := claimer.BuildPresentation(issuerPubKey, credential, request)
	if err != nil {
		return nil, err
	}
	return disclosedAttr, nil
}

func BuildCombinedPresentation(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 4 {
		return nil, errors.New("missing inputs to build combined presentation")
	}
	claimer := &credentials.Claimer{}
	creds := []*credentials.AttestedClaim{}
	request := &credentials.CombinedPresentationRequest{}
	attesterPubKey := []*gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), &creds); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), request); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[3].String()), &attesterPubKey); err != nil {
		return nil, err
	}

	disclosedAttr, err := claimer.BuildCombinedPresentation(attesterPubKey, creds, request)
	if err != nil {
		return nil, err
	}
	return disclosedAttr, nil
}

// UpdateCredential updates the non revocation witness using the provided update.
func UpdateCredential(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 3 {
		return nil, errors.New("missing inputs to update credential")
	}
	credential := &credentials.AttestedClaim{}
	update := &revocation.Update{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), credential); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), update); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), issuerPubKey); err != nil {
		return nil, err
	}

	if err := credential.Update(issuerPubKey, update); err != nil {
		return nil, err
	}
	return credential, nil
}

// UpdateAllCredential updates the non revocation witness using all the provided updates.
func UpdateAllCredential(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 3 {
		return nil, errors.New("missing inputs to update credential")
	}
	credential := &credentials.AttestedClaim{}
	updates := []*revocation.Update{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), credential); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), &updates); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), issuerPubKey); err != nil {
		return nil, err
	}

	if err := credential.UpdateAll(issuerPubKey, updates); err != nil {
		return nil, err
	}
	return credential, nil
}

File: src/claim/Claimer.ts
import toSeed from '@polkadot/util-crypto/mnemonic/toSeed'
import { u8aToHex } from '@polkadot/util'
import validate from '@polkadot/util-crypto/mnemonic/validate'
import IClaimer, {
  AttestationRequest,
  ClaimerAttestationSession,
  Presentation,
  CombinedPresentation,
  ClaimError,
} from '../types/Claim'
import WasmHooks from '../wasm/WasmHooks'
import {
  IGabiMsgSession,
  InitiateAttestationRequest,
  Attestation,
  AttesterPublicKey,
  KeyLength,
  DEFAULT_KEY_LENGTH,
} from '../types/Attestation'
import {
  CombinedPresentationRequest,
  PresentationRequest,
} from '../types/Verification'
import goWasmExec from '../wasm/wasm_exec_wrapper'
import Credential from './Credential'

/**
 * Checks that the provided claim is a valid object.
 *
 * @param claim The object which should be a valid claim.
 * @throws [[ClaimError.claimMissing]] If the claim is missing inside the [[AttestationRequest]].
 * @throws [[ClaimError.notAnObject]] If the [[Attestation]] object includes a non-object type claim.
 * @throws [[ClaimError.duringParsing]] If an error occurs during JSON deserialization.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function checkValidClaimStructure(claim: Record<string, any>): void | Error {
  if (!Object.keys(claim).length) {
    throw ClaimError.claimMissing
  }
  if (typeof claim !== 'object') {
    throw ClaimError.notAnObject(typeof claim)
  }
  if (Array.isArray(claim)) {
    throw ClaimError.duringParsing
  }
}

export default class Claimer implements IClaimer {
  private readonly secret: string

  /**
   * Generates a claimer using the provided mnemonic.
   *
   * @param mnemonic The mnemonic which is used to generate the key.
   * @param options An optional object containing options for the key generation.
   * @param options.password The password which is used to generate the key.
   * @param options.keyLength The key length of the new secret. Note that this secret will only support credentials and attester with the same key length.
   * @returns A new [[Claimer]].
   */
  public static async buildFromMnemonic(
    mnemonic: string,
    {
      password,
      keyLength,
    }: {
      password?: string
      keyLength?: KeyLength
    } = {}
  ): Promise<Claimer> {
    if (!validate(mnemonic)) {
      throw new Error('Invalid mnemonic')
    }
    const seed = toSeed(mnemonic, password)
    return this.buildFromSeed(seed, keyLength)
  }

  /**
   * Generates a claimer using the provided seed.
   *
   * @param seed The seed which is used to generate the key.
   * @param keyLength The key length of the new secret. Note that this secret will only support credentials and attester with the same key length.
   * @returns A new [[Claimer]].
   */
  public static async buildFromSeed(
    seed: Uint8Array,
    keyLength?: KeyLength
  ): Promise<Claimer> {
    // secret's structure unmarshalled is { MasterSecret: string }
    const secret = await goWasmExec<string>(WasmHooks.keyFromSeed, [
      u8aToHex(seed),
      keyLength || DEFAULT_KEY_LENGTH,
    ])
    return new this(secret)
  }

  /**
   * Generates a new master secret and returns a new [[Claimer]] object.
   *
   * @returns A new [[Claimer]].
   */
  public static async create(): Promise<Claimer> {
    const secret = await goWasmExec<string>(WasmHooks.genKey)
    return new this(secret)
  }

  /**
   * Constructs a new [[Claimer]] using the given master secret.
   *
   * @param secret The master secret of the [[Claimer]].
   */
  constructor(secret: string) {
    this.secret = secret
  }

  /**
   * Creates an [[AttestationRequest]] using the provided [[InitiateAttestationRequest]].
   *
   * @param p The parameter object.
   * @param p.claim The claim which should get attested.
   * @param p.startAttestationMsg The [[InitiateAttestationRequest]] provided by the attester.
   * @param p.attesterPubKey The [[AttesterPublicKey]].
   * @returns An [[AttestationRequest]] and a [[ClaimerAttestationSession]] which together with an [[Attestation]]
   * can be used to create a [[Credential]].
   */
  public async requestAttestation({
    claim,
    startAttestationMsg,
    attesterPubKey,
  }: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    claim: Record<string, any>
    startAttestationMsg: InitiateAttestationRequest
    attesterPubKey: AttesterPublicKey
  }): ReturnType<IClaimer['requestAttestation']> {
    // check for invalid claim structure
    checkValidClaimStructure(claim)
    const { message, session } = await goWasmExec<IGabiMsgSession>(
      WasmHooks.requestAttestation,
      [
        this.secret,
        JSON.stringify(claim),
        startAttestationMsg.toString(),
        attesterPubKey.toString(),
      ]
    )
    return {
      message: new AttestationRequest(message),
      session: new ClaimerAttestationSession(session),
    }
  }

  /**
   * Builds a [[Credential]] using the [[ClaimerAttestationSession]] and the [[Attestation]].
   *
   * @param p The parameter object.
   * @param p.claimerSession The session object corresponding to the [[AttestationRequest]].
   * @param p.attestation The [[Attestation]] provided by the [[Attester]].
   * @returns A signed and valid [[Credential]].
   */
  public async buildCredential({
    claimerSession,
    attestation,
  }: {
    claimerSession: ClaimerAttestationSession
    attestation: Attestation
  }): ReturnType<IClaimer['buildCredential']> {
    return new Credential(
      await goWasmExec<string>(WasmHooks.buildCredential, [
        this.secret,
        claimerSession.toString(),
        attestation.toString(),
      ])
    )
  }

  /**
   * Uses the [[PresentationRequest]] and a [[Credential]] to build an anonymous presentation.
   *
   * @param p The parameter object.
   * @param p.credential The [[Credential]] which contains all the requested attributes.
   * @param p.presentationReq The [[PresentationRequest]] received from the Verifier.
   * @param p.attesterPubKey The public key of the [[Attester]] who signed the [[Credential]].
   * @returns A [[Presentation]] that can be used to disclose attributes with a Verifier.
   *    Must only be used once!
   */
  public async buildPresentation({
    credential,
    presentationReq,
    attesterPubKey,
  }: {
    credential: Credential
    presentationReq: PresentationRequest
    attesterPubKey: AttesterPublicKey
  }): ReturnType<IClaimer['buildPresentation']> {
    return new Presentation(
      await goWasmExec<string>(WasmHooks.buildPresentation, [
        this.secret,
        credential.toString(),
        presentationReq.toString(),
        attesterPubKey.toString(),
      ])
    )
  }

  /**
   * Uses the [[PresentationRequest]] and a [[Credential]] to build an anonymous presentation.
   *
   * @param p The parameter object.
   * @param p.credentials An array of [[Credential]]s which is used to provide the requested attributes.
   * @param p.combinedPresentationReq The array of [[PresentationRequest]]s received from the Verifier.
   * @param p.attesterPubKeys An array of [[AttesterPublicKey]]s which corresponds to the array of [[Credential]]s.
   * @returns A [[CombinedPresentation]] that can be used to disclose attributes with a Verifier.
   *    Must only be used once!
   */
  public async buildCombinedPresentation({
    credentials,
    combinedPresentationReq,
    attesterPubKeys,
  }: {
    credentials: Credential[]
    combinedPresentationReq: CombinedPresentationRequest
    attesterPubKeys: AttesterPublicKey[]
  }): ReturnType<IClaimer['buildCombinedPresentation']> {
    // make a json array out of already json serialised values
    // we don't want a json array of strings
    return new CombinedPresentation(
      await goWasmExec<string>(WasmHooks.buildCombinedPresentation, [
        this.secret,
        `[${credentials.join(',')}]`,
        combinedPresentationReq.toString(),
        `[${attesterPubKeys.join(',')}]`,
      ])
    )
  }
}
